### ISSUE TITLE
ty ignores __class_getitem__ since latest version?

### ISSUE THREAD

**Issue by Holt59** [astral-sh/ty#4331]
### Summary

The following codes was working with `ty` 0.0.65 but is not working with `ty` 0.0.66 (and above):

```python
from __future__ import annotations

class U:
    def __class_getitem__(self, value: int) -> type[U]:
        return U

x: U[0]  # Cannot subscript non-generic type `<class 'U'>` (not-subscriptable) [Ln 9, Col 4]
```

I don't think that this is expected.

**Comment by carljm** [comment]
Hi, thanks for the report! This is a bit subtle.

`U[0]` is not a valid type expression in the Python type system, because `U` is not defined as a generic class. A custom definition of `__class_getitem__` does not make it a generic class, or make `U[0]` valid as a type expression. Type expressions have a defined grammar and semantics -- arbitrary Python expressions evaluating to `type[...]` are not automatically valid type expressions.

That said, the diagnostic code we should emit here is `invalid-type-form`, rather than `not-subscriptable`. The latter is clearly wrong, `U` is subscriptable at runtime.

If you are using type annotations in your code for some non-typing purpose, you'd want to disable the `invalid-type-form` error code.

**Comment by Holt59** [comment]
Any reason why this was working prior to `ty` 0.0.66?

**Comment by carljm** [comment]
It wasn't "working" in earlier versions of ty (at least not in the sense of ty actually treating `U[0]` as an annotation equivalent to `U`) -- we just hadn't implemented that particular validation yet, so it was treating it as a dynamic `@Todo` type, equivalent to `Any`.

**Comment by Holt59** [comment]
> If you are using type annotations in your code for some non-typing purpose, you'd want to disable the invalid-type-form error code.

Is there a way to disable this for only a single class? Or a way to change the stub of the class to not generate such error?

My current use-case is creating stubs for `astropy` type annotation (https://docs.astropy.org/en/latest/units/type_hints.html#preserving-units), so basically `Quantity[km]` where `km` is not a type but an instance of a class.

**Comment by carljm** [comment]
The workarounds would be to annotate as just `Quantity` with no unit, or to use `Annotated` explicitly (`Annotated[Quantity, Unit("m")]`).

Given that astropy is explicitly encouraging users to do this, we may want to consider supporting it, but I think it should be clearly specified as either supported or unsupported. I opened a discussion at https://discuss.python.org/t/should-non-generic-classes-with-class-getitem-be-subscriptable-in-a-type-expression/108666
